nomad/grapl-generator.nomad

```diff
@@ -0,0 +1,72 @@
+variable "tenant_id" {
+    type = string
+    description = "The opaque tenant id."
+}
+
+variable "generator_ingest" {
+    type = string
+    description = "The expected data ingest type. e.g. sysmon."
+}
+
+variable "generator_artifact_url" {
+    type = string
+    description = "The url of the generator artifact to run. Supported formats: http(s), git, hg, and s3."
+}
+
+variable "generator_count" {
+    type = number
+    default = 1
+}
+
+job "grapl-tenant-generator" {
+    namespace = "generator-${var.tenant_id}-${var.generator_ingest}"
+    datacenters = ["dc1"]
+
+    type = "service"
+
+    update {
+        # Automatically promotes to canaries if all canaries are healthy during an update / deployment 
+        auto_promote = true
+        # Auto reverts to the last stable job variant if the update fails 
+        auto_revert = true
+        # Spins up a "canary" instance of potentially destructive updates, validates that they are healthy, then promotes the instance to update
+        canary = 1
+        max_parallel = 1
+        # The min amount of reported "healthy" time before a instance is considered healthy and an allocation is opened up for further updates
+        min_healthy_time = "15s"
+    }
+
+    group "generator" {
+        network {
+            port "generator_receiver" { }
+        }
+
+        count = var.generator_count
+
+        task "tenant-generator" {
+            driver = "exec"
+
+            artifact {
+                source = var.generator_artifact_url
+                destination = "local/generator"
+                mode = "file"
+            }
+
+            env {
+                TENANT_ID = "${var.tenant_id}"
+                INGEST_TYPE = "${var.generator_ingest}"
+            }
+
+            config {
+                command = "generator"
+            }
+
+            service {
+                name = "generator-${var.tenant_id}-${var.generator_ingest}"
+                port = "generator_receiver"
+                tags = ["generator-${var.tenant_id}"]
+                canary_tags = ["canary-generator-${var.tenant_id}"]
```

Comment:
Should this not incorporate the `generator_ingest` as well (same for `tags`, too)?

Reply:
The intent here was that if we wanted to easily identify all of the generators for a particular tenant, we can open the consul interface and just select the `generator-<tenant_id>` tag. We could also add a `ingest-<generator_ingest>` tag if we want to find all of the generators that support a particular ingest type too!

Do you think that'd be useful?

Reply:
I haven't messed around with Nomad's tagging yet; can you search on multiple tags at once? If so, then it might just be better to tag by each individual bit of data and let the person doing the search compose them as they need, rather than us having to guess beforehand what kinds of searches we're likely to want to do.

Are tags just a single value, or can they be a key/value pair, like in AWS?

Reply:
> I haven't messed around with Nomad's tagging yet; can you search on multiple tags at once? 

Just to clear up confusion, the tags on a `service` stanza are for Consul, not Nomad. When you use the Consul interface, it'll let you filter down the services more easily by searching using the search bar. You also don't need the full name, either. You could search for `canary` and all of the `canary-generator-<tenant_id>` tagged instances would should up.

> If so, then it might just be better to tag by each individual bit of data and let the person doing the search compose

I assume you mean tag with `tenant_id` and `generator_ingest` individually? I could do that!

> Are tags just a single value...?

Single value, yea. There is a metadata field you can add on though, if you want that. I don't know if metadata can be searched but I suspect no.

Reply:
Gotcha; makes sense re: Consul vs. Nomad.

Yeah, I think going with the more atomic tags makes more sense at this point. We can `AND` things together when we need to. Ultimately, I think this will end up being more flexible.

Reply:
As for the single vs K/V question, I was thinking it'd be nice if we could say something like `tenant_id: BLAH`, but using `tenant_id_${BLAH}` is fine (particularly if that's our only option 😅)